aggregate_spatial_window.json

```
@@ -0,0 +1,95 @@
+{
+  "id": "aggregate_spatial_window",
+  "summary": "Zonal statistics for rectangular windows",
+  "description": "Aggregates statistics over the spatial dimensions. This process passes a list of values to the reducer.\n\nThe results are computed dividing the input X/Y grid into x_size/y_size non-overlapping windows and applying the reducer process to each window.\nIf X and Y are not multiple of x_size/y_size NaNs padding before the computation is applied.",
+  "categories": [
+    "cubes",
+    "aggregate & resample"
+  ],
+  "parameters": [
+    {
+      "name": "data",
+      "description": "A raster data cube.",
+      "schema": {
+        "type": "object",
+        "subtype": "raster-cube"
+      }
+    },
+    {
+      "name": "reducer",
+      "description": "A reducer to be applied on all values of each geometry. A reducer is a single process such as ``mean()`` or a set of processes, which computes a single value for a list of values, see the category 'reducer' for such processes.",
+      "schema": {
+        "type": "object",
+        "subtype": "process-graph",
+        "parameters": [
+          {
+            "name": "data",
+            "description": "An array with elements of any type.",
+            "schema": {
+              "type": "array",
+              "items": {
+                "description": "Any data type."
+              }
+            }
+          },
+          {
+            "name": "context",
+            "description": "Additional data passed by the user.",
+            "schema": {
+              "description": "Any data type."
+            },
+            "optional": true,
+            "default": null
+          }
+        ]
+      }
+    },
+    {
+      "name": "x_size",
+      "description": "Window size along X or Longitude. The number of values used as input to the reducer along X dimension.",
+      "schema": {
+        "type": "number"
+      },
+      "optional": true,
+      "default": 1
+    },
+    {
+      "name": "y_size",
+      "description": "Window size along Y or Latitude. The number of values used as input to the reducer along Y dimension.",
+      "schema": {
+        "type": "number"
+      },
+      "optional": true,
+      "default": 1
+    },
+    {
+      "name": "boundary",
+      "description": "Operation to apply if X/Y are not multiple of x_size/y_size. Options are:\n\n- pad (default): pad the datacube with NaNs to fit window dimensions.\n\n- trim: trim the datacube to fit window dimensions.",
```

```suggestion
      "description": "Behavior to apply if the number of values for the axes `x` and `y` is not a multiple of the corresponding `x_size` or `y_size` parameter. Options are:\n\n- `pad` (default): pad the data cube with `NaN` (not a number) values to fit the required window size.\n\n- `trim`: trim the data cube to fit the required window size.",
```

Open questions: What is trimmed? Equally on each side? At the "end"? At the "beginning"?

---

Good question. Actually I'm using xarray coarsen method, which has also the option to align top-left or top-right the windows http://xarray.pydata.org/en/v0.15.1/generated/xarray.DataArray.coarsen.html
So, if aligned to the left it will trim the excess on the right and vice versa.

---

resample_spatial has the align parameter, which is very similar.

---

I updated the description as you have suggested, we still need to decide what to do for the alignment.

---

Yes, I'm unsure yet. I'll put this into proposal stage first (see #196) and then we can elaborate during development whether a paramter should be added. For now we should just specify a default, either upper-left or lower-right. I think I'd default to pad the lower-right. By the way, why do we fill with NaN instead of null/no-data? That seems the approach more aligned with openEO. NaN is actually rarely used here...

---

I was reporting what the Xarray library does, but we are free to choose something else.

---

I think no-data / null makes more sense in the context of openEO. no data is implemented into several processes, but NaN is not (really).

---

Ok I'll put no-data instead of Nan.